python/tests/test_topology.py

```diff
@@ -168,6 +168,103 @@ def kc_distance(tree1, tree2, lambda_=0):
+    for j, u in enumerate(tree.tree_sequence.samples()):
+        sample_index_map[u] = j
+    for root in tree.roots:
+        stack = [(tree.root, 0)]
+        while len(stack) > 0:
+            u, depth = stack.pop()
+            if tree.is_leaf(u):
+                time = 0 if u is root else tree.branch_length(u)
+                update_kc_vectors_single_leaf(ts, kc_vecs, u, time, sample_index_map)
+            else:
+                c1 = tree.left_child(u)
+                while c1 != tskit.NULL:
+                    stack.append((c1, depth + 1))
+                    c2 = tree.right_sib(c1)
+                    while c2 != tskit.NULL:
+                        update_kc_vectors_all_pairs(
+                            tree,
+                            kc_vecs,
+                            c1,
+                            c2,
+                            depth,
+                            tree.time(root) - tree.time(u),
+                            sample_index_map,
+                        )
+                        c2 = tree.right_sib(c2)
+                    c1 = tree.right_sib(c1)
+
+
+def update_kc_vectors_single_leaf(ts, kc_vecs, u, time, sample_index_map):
+    u_index = int(sample_index_map[u])
+    kc_vecs.m[kc_vecs.N + u_index] = 1
+    kc_vecs.M[kc_vecs.N + u_index] = time
+
+
+def update_kc_vectors_all_pairs(tree, kc_vecs, c1, c2, depth, time, sample_index_map):
+    leaves = tree.tree_sequence.samples()
+    c1_left = tree.left_sample(c1)
+    c1_right = tree.right_sample(c1)
+    leaf1_index = c1_left
+    while True:
+        c2_left = tree.left_sample(c2)
+        c2_right = tree.right_sample(c2)
+        leaf2_index = c2_left
+        while True:
+            leaf1 = leaves[leaf1_index]
+            leaf2 = leaves[leaf2_index]
+            update_kc_vectors_pair(kc_vecs, leaf1, leaf2, depth, time, sample_index_map)
+            if leaf2_index == c2_right:
+                break
+            leaf2_index = tree.next_sample(leaf2_index)
+        if leaf1_index == c1_right:
+            break
+        leaf1_index = tree.next_sample(leaf1_index)
+
+
+def update_kc_vectors_pair(kc_vecs, u, v, depth, time, sample_index_map):
+    n1 = int(min(sample_index_map[u], sample_index_map[v]))
+    n2 = int(max(sample_index_map[u], sample_index_map[v]))
+    pair_index = n2 - n1 - 1 + (-1 * n1 * (n1 - 2 * kc_vecs.n + 1)) // 2
+
+    kc_vecs.m[pair_index] = depth
+    kc_vecs.M[pair_index] = time
+
+
+def norm_kc_vectors(kc_vecs1, kc_vecs2, lambda_):
+    vT1 = 0
+    vT2 = 0
+    distance_sum = 0
+    for i in range(kc_vecs1.n + kc_vecs1.N):
+        vT1 = (kc_vecs1.m[i] * (1 - lambda_)) + (lambda_ * kc_vecs1.M[i])
+        vT2 = (kc_vecs2.m[i] * (1 - lambda_)) + (lambda_ * kc_vecs2.M[i])
+        distance_sum += (vT1 - vT2) ** 2
+
+    return math.sqrt(distance_sum)
+
+
 def kc_distance_simple(tree1, tree2, lambda_=0):
     """
     Simplified version of the kc_distance() function above.
```

Comment:
I think we should keep this version of kc_distance and replace the other one above. The idea is that we have a very naive, literal implementation of the metric as a base to check our more sophisticated algorithm against.  This version looks a bit simpler than the one above, right?

Reply:
I agree that the most straight-forward way to think about it is:
Cartesian product of leaves --> mrca --> depth/time
I think I could write that even a bit simpler than the two-step process that's being done here.

On a related note, I find the name `kc_distance_simple` to be kind of confusing, because I don't know whether to associate simple with naive or without advanced features. How about bare_bones or c_compatible?

Reply:
Sure - how about ``naive_kc_distance``?